tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,681 @@
+        loop {
+            // Has the semaphore closed?
+            if curr & CLOSED > 0 {
+                return Err(TryAcquireError::Closed);
+            }
+
+            // Are there enough permits remaining?
+            if (curr as u16) < num_permits {
+                return Err(TryAcquireError::NoPermits);
+            }
+
+            let next = curr - num_permits as usize;
+
+            match self.permits.compare_exchange(curr, next, AcqRel, Acquire) {
+                Ok(_) => return Ok(()),
+                Err(actual) => curr = actual,
+            }
+        }
+    }
+
+    fn poll_acquire(
+        &self,
+        cx: &mut Context<'_>,
+        needed: u16,
+        node: Pin<&mut Waiter>,
+        queued: bool,
+    ) -> Poll<Result<(), AcquireError>> {
+        let mut acquired = 0;
+
+        // If we are already in the wait queue, we need to lock the queue so we
+        // can access the wait queue entry's current state.
+        let (needed, mut lock) = if queued {
+            let lock = self.waiters.lock().unwrap();
+            // Safety: since we have acquired the lock, it is safe to look at
+            // the waiter's state.
+            let needed = node.state.with(|curr| unsafe { *curr });
+            (needed, Some(lock))
+        } else {
+            (needed as usize, None)
+        };
+
+        // First, try to take the requested number of permits from the
+        // semaphore.
+        let mut curr = self.permits.load(Acquire);
+        let mut waiters = loop {
+            // Has the semaphore closed?
+            if curr & CLOSED > 0 {
+                return Ready(Err(AcquireError::closed()));
+            }
+
+            let mut remaining = 0;
+            let (next, acq) = if curr + acquired >= needed {
+                let next = curr - (needed - acquired);
+                (next, needed)
+            } else {
+                remaining = (needed - acquired) - curr;
+                (0, curr)
+            };
+
+            if remaining > 0 && lock.is_none() {
+                // No permits were immediately available, so this permit will
+                // (probably) need to wait. We'll need to acquire a lock on the
+                // wait queue.
+                if let Ok(l) = self.waiters.try_lock() {
+                    // If we were able to acquire the lock *now* without
+                    // blocking, we don't need to reload the semaphore's permit state.
+                    lock = Some(l)
+                } else {
+                    lock = Some(self.waiters.lock().unwrap());
+                    // While we were waiting to lock the wait list, additional
+                    // permits may have been released. Therefore, we will acquire a
+                    // new snapshot of the current state of the semaphore before
+                    // actually enqueueing the waiter..
+                    curr = self.permits.load(Acquire);
+                    continue;
+                }
```

Comment:
I'm not following the reasoning for this logic vs. just acquire the lock and attempting the compare-exchange.

Reply:
Ah, I think this is not necessary when the waiter state is protected by the lock. 

When the waiter's state was stored in the atomic, this was necessary to ensure that our view of the waiter state atomic was kept up to date --- permits may have been assigned while we were waiting for the lock, but the CAS on the _semaphore_'s permit counter could succeed if all the permits were assigned to the wait list.

This can probably be removed if we don't make the waiter state atomic, but if we want to bring that back, I think it will still be necessary.

Reply:
I don't understand why it would be necessary for the waiter state atomic? My assumption is that once the num of permits is requested, the waiter will not touch that value until the permits are assigned. The waiter may drop, which requires the lock anyway.

I know that in the **old** semaphore, there was shenanigans to handle the number of permits *changing* in-flight, but that should be removed IMO.